File: algo/src/main/java/com/hartwig/actin/algo/evaluation/tumor/HasSolidPrimaryTumorIncludingLymphoma.java
package com.hartwig.actin.algo.evaluation.tumor;

import java.util.Set;

import com.google.common.collect.Sets;
import com.hartwig.actin.PatientRecord;
import com.hartwig.actin.algo.datamodel.Evaluation;
import com.hartwig.actin.algo.datamodel.EvaluationResult;
import com.hartwig.actin.algo.datamodel.ImmutableEvaluation;
import com.hartwig.actin.algo.doid.DoidModel;
import com.hartwig.actin.algo.evaluation.EvaluationFactory;
import com.hartwig.actin.algo.evaluation.EvaluationFunction;

import org.jetbrains.annotations.NotNull;

public class HasSolidPrimaryTumorIncludingLymphoma implements EvaluationFunction {

    static final String CANCER_DOID = "162";
    static final Set<String> NON_SOLID_CANCER_DOIDS = Sets.newHashSet();
    static final Set<String> WARN_SOLID_CANCER_DOIDS = Sets.newHashSet();

    static {
        NON_SOLID_CANCER_DOIDS.add("1240"); // leukemia
        NON_SOLID_CANCER_DOIDS.add("712"); // refractory hematologic cancer
        NON_SOLID_CANCER_DOIDS.add("4960"); // bone marrow cancer

        WARN_SOLID_CANCER_DOIDS.add("5772"); // central nervous system hematologic cancer
        WARN_SOLID_CANCER_DOIDS.add("3282"); // dendritic cell thymoma
        WARN_SOLID_CANCER_DOIDS.add("5621"); // histiocytic and denditric cell cancer
        WARN_SOLID_CANCER_DOIDS.add("3664"); // mast cell neoplasm
        WARN_SOLID_CANCER_DOIDS.add("8683"); // myeloid sarcoma
    }

    @NotNull
    private final DoidModel doidModel;

    HasSolidPrimaryTumorIncludingLymphoma(@NotNull final DoidModel doidModel) {
        this.doidModel = doidModel;
    }

    @NotNull
    @Override
    public Evaluation evaluate(@NotNull PatientRecord record) {
        Set<String> doids = record.clinical().tumor().doids();

        if (doids == null || doids.isEmpty()) {
            return EvaluationFactory.unrecoverable()
                    .result(EvaluationResult.UNDETERMINED)
                    .addUndeterminedSpecificMessages(
                            "No tumor location/type configured for patient, unknown if solid primary tumor or lymphoma")
                    .addUndeterminedGeneralMessages("Unconfigured tumor location/type")
                    .build();
        }

        EvaluationResult result =
                DoidEvaluationFunctions.evaluate(doidModel, doids, CANCER_DOID, NON_SOLID_CANCER_DOIDS, WARN_SOLID_CANCER_DOIDS);

        ImmutableEvaluation.Builder builder = EvaluationFactory.unrecoverable().result(result);
        if (result == EvaluationResult.FAIL) {
            builder.addFailSpecificMessages("Patient has non-solid primary tumor");
            builder.addFailGeneralMessages("Tumor type");
        } else if (result == EvaluationResult.WARN) {
            builder.addWarnSpecificMessages("Patient has potentially non-solid primary tumor");
            builder.addWarnGeneralMessages("Tumor type");
        } else if (result == EvaluationResult.PASS) {
            builder.addPassSpecificMessages("Patient has solid primary tumor or lymphoma");
            builder.addPassGeneralMessages("Tumor type");
        }

        return builder.build();
    }
}

File: algo/src/main/kotlin/com/hartwig/actin/algo/evaluation/tumor/HasSolidPrimaryTumorIncludingLymphoma.kt
package com.hartwig.actin.algo.evaluation.tumor

import com.hartwig.actin.algo.doid.DoidConstants
import com.hartwig.actin.algo.evaluation.EvaluationFactory
import com.hartwig.actin.algo.evaluation.EvaluationFunction
import com.hartwig.actin.algo.evaluation.EvaluationLabels
import com.hartwig.actin.datamodel.PatientRecord
import com.hartwig.actin.datamodel.algo.Evaluation
import com.hartwig.actin.datamodel.algo.EvaluationResult
import com.hartwig.actin.doid.DoidModel

class HasSolidPrimaryTumorIncludingLymphoma(private val doidModel: DoidModel, private val labels: EvaluationLabels.Tumor) :
    EvaluationFunction {

    override fun evaluate(record: PatientRecord): Evaluation {
        val tumorDoids = record.tumor.doids
        if (!DoidEvaluationFunctions.hasConfiguredDoids(tumorDoids)) {
            return EvaluationFactory.undetermined(labels.hasSolidPrimaryTumorIncludingLymphomaUndetermined())
        }
        val result = DoidEvaluationFunctions.evaluateAllDoidsMatchWithFailAndWarns(
            doidModel,
            tumorDoids,
            setOf(DoidConstants.CANCER_DOID, DoidConstants.BENIGN_NEOPLASM_DOID),
            DoidConstants.NON_SOLID_CANCER_DOIDS,
            WARN_SOLID_CANCER_DOIDS
        )
        return when (result) {
            EvaluationResult.FAIL -> {
                EvaluationFactory.fail(labels.hasSolidPrimaryTumorIncludingLymphomaFail())
            }

            EvaluationResult.WARN -> {
                EvaluationFactory.warn(labels.hasSolidPrimaryTumorIncludingLymphomaWarn())
            }

            EvaluationResult.PASS -> {
                EvaluationFactory.pass(labels.hasSolidPrimaryTumorIncludingLymphomaPass())
            }

            else -> {
                Evaluation(result = result, recoverable = false)
            }
        }
    }

    companion object {
        val WARN_SOLID_CANCER_DOIDS = setOf(
            DoidConstants.CENTRAL_NERVOUS_SYSTEM_HEMATOLOGIC_CANCER_DOID,
            DoidConstants.DENDRITIC_CELL_THYMOMA_DOID,
            DoidConstants.HISTIOCYTIC_AND_DENDRITIC_CELL_CANCER_DOID,
            DoidConstants.MAST_CELL_NEOPLASM_DOID,
            DoidConstants.MYELOID_SARCOMA_DOID,
        )
    }
}